src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/UpToDate/UpToDateCheckImplicitConfiguredInput.cs

```diff
@@ -0,0 +1,572 @@
+            string? copyUpToDateMarkerItem,
+            string? outputRelativeOrFullPath,
+            string? newestImportInput,
+            IComparable? lastVersionSeen,
+            bool isDisabled,
+            ImmutableHashSet<string> itemTypes,
+            ImmutableDictionary<string, ImmutableHashSet<(string, string?, BuildUpToDateCheck.CopyType)>> itemsByItemType,
+            ImmutableDictionary<string, ImmutableHashSet<string>> upToDateCheckInputItemsBySetName,
+            ImmutableDictionary<string, ImmutableHashSet<string>> upToDateCheckOutputItemsBySetName,
+            ImmutableDictionary<string, ImmutableHashSet<string>> upToDateCheckBuiltItemsBySetName,
+            ImmutableDictionary<string, string> copiedOutputFiles,
+            ImmutableHashSet<string> resolvedAnalyzerReferencePaths,
+            ImmutableHashSet<string> resolvedCompilationReferencePaths,
+            ImmutableHashSet<string> copyReferenceInputs,
+            IImmutableDictionary<string, DateTime> additionalDependentFileTimes,
+            DateTime lastAdditionalDependentFileTimesChangedAtUtc,
+            DateTime lastItemsChangedAtUtc,
+            ImmutableArray<(bool IsAdd, string ItemType, string Path, string? Link, BuildUpToDateCheck.CopyType)> lastItemChanges)
+        {
+            MSBuildProjectFullPath = msBuildProjectFullPath;
+            MSBuildProjectDirectory = msBuildProjectDirectory;
+            CopyUpToDateMarkerItem = copyUpToDateMarkerItem;
+            OutputRelativeOrFullPath = outputRelativeOrFullPath;
+            NewestImportInput = newestImportInput;
+            LastVersionSeen = lastVersionSeen;
+            IsDisabled = isDisabled;
+            ItemTypes = itemTypes;
+            ItemsByItemType = itemsByItemType;
+            UpToDateCheckInputItemsBySetName = upToDateCheckInputItemsBySetName;
+            UpToDateCheckOutputItemsBySetName = upToDateCheckOutputItemsBySetName;
+            UpToDateCheckBuiltItemsBySetName = upToDateCheckBuiltItemsBySetName;
+            CopiedOutputFiles = copiedOutputFiles;
+            ResolvedAnalyzerReferencePaths = resolvedAnalyzerReferencePaths;
+            ResolvedCompilationReferencePaths = resolvedCompilationReferencePaths;
+            CopyReferenceInputs = copyReferenceInputs;
+            LastItemsChangedAtUtc = lastItemsChangedAtUtc;
+            AdditionalDependentFileTimes = additionalDependentFileTimes;
+            LastAdditionalDependentFileTimesChangedAtUtc = lastAdditionalDependentFileTimesChangedAtUtc;
+            LastItemChanges = lastItemChanges;
+
+            var setNames = new HashSet<string>(BuildUpToDateCheck.SetNameComparer);
+            AddKeys(upToDateCheckInputItemsBySetName);
+            AddKeys(upToDateCheckOutputItemsBySetName);
+            AddKeys(upToDateCheckBuiltItemsBySetName);
+            setNames.Remove(BuildUpToDateCheck.DefaultSetName);
+            SetNames = setNames.OrderBy(n => n, BuildUpToDateCheck.SetNameComparer).ToImmutableArray();
```

Comment:
This could be accomplished in 1 Linq expression. Not sure if it is any faster or better. (Note: I guess there isn't decontruction support in lambdas yet: https://github.com/dotnet/csharplang/discussions/258)
```suggestion
SetNames = upToDateCheckInputItemsBySetName.Concat(upToDateCheckOutputItemsBySetName).Concat(upToDateCheckBuiltItemsBySetName).Select(x => x.Key).Distinct().Where(n => n != BuildUpToDateCheck.DefaultSetName).OrderBy(n => n, BuildUpToDateCheck.SetNameComparer).ToImmutableArray();
```

Reply:
062731b7f9540e7441e49f9a23e6c0c6f5986630 explicitly unrolled this to avoid allocations. Linq operates on `IEnumerable<>` which will always allocate an `IEnumerator<>` on the heap. By using `foreach` on the concrete collection type in the `AddKeys` method, the compiler will use the struct enumerator which can be kept on the stack.

> I guess there isn't decontruction support in lambdas yet

I would love this. Also the ability to deconstruct in method parameter lists:

```c#
public void M((int i, string s))
{
}
```